servant-server/test/Servant/ServerSpec.hs

```diff
@@ -728,3 +731,80 @@ miscReqCombinatorsSpec = with (return $ serve miscApi miscServ) $
       go "/host" "\"0.0.0.0:0\""
 
   where go path res = Test.Hspec.Wai.get path `shouldRespondWith` res
+
+
+type ErrorRoutingUser404
+  =    "un"   :> Get '[JSON] Int
+  :<|> "deux" :> Get '[JSON] Int
+  :<|>           Get '[JSON] Int
+
+errorRoutingUser404Spec :: Spec
+errorRoutingUser404Spec =
+  describe "Handlers can trigger re-routing by responding with error status 404" $ do
+    describe "happy handlers" $ do
+      let hs = return 1 :<|> return 2 :<|> return 3
+      go hs "/un"   "1"
+      go hs "/deux" "2"
+      go hs "/"     "3"
+      go hs "/n/a"  404
+    describe "re-route from /un handler" $ do
+      let hs = left err404 :<|> return 2 :<|> return 3
+      go hs "/un"   "2"
+      go hs "/deux" "2"
+      go hs "/"     "3"
+      go hs "/n/a"  404
+  where
+    go :: Server ErrorRoutingUser404 -> String -> ResponseMatcher -> Spec
+    go hs path resp = with (return $ serve (Proxy :: Proxy ErrorRoutingUser404) hs) $
+      it path $ Test.Hspec.Wai.get (cs path) `shouldRespondWith` resp
+
+
+type ErrorRoutingBodyParseError
+  =    "sum" :> ReqBody '[JSON] [Int] :> Post '[JSON] Int
+  :<|> "const"                        :> Post '[JSON] Int
+  :<|> Raw
+
+errorRoutingBodyParseErrorSpec :: Spec
+errorRoutingBodyParseErrorSpec =
+  describe "Broken request body triggers error response (not re-routing)" $ do
+    let hs = (return . sum) :<|> return 2 :<|> (\_ cont -> cont (responseLBS ok200 [] "Raw"))
+    describe "happy handlers" $ do
+      go hs "/sum"   "[1, 2]" "3"
```

Comment:
The reason this is failing is because the `Content-Type` header is not set, so we don't know how to decode it. There's an issue for trying to decode as the first content-type in the CT list ( #198 ), but I think we shouldn't try to do all of that in one go.